core/peerset/src/lib.rs

```diff
@@ -147,156 +130,176 @@ pub struct PeersetConfig {
 }
 
 impl Peerset {
 	/// Builds a new peerset from the given configuration.
 	pub fn from_config(config: PeersetConfig) -> (Peerset, PeersetHandle) {
 		let (tx, rx) = mpsc::unbounded();
 
-		let data = PeersetData {
-			discovered: Slots::new(DISCOVERED_NODES_LIMIT),
-			reserved_only: config.reserved_only,
-			out_slots: Slots::new(config.out_peers),
-			in_slots: Slots::new(config.in_peers),
-			scores: LruCache::new(PEERSET_SCORES_CACHE_SIZE),
-		};
-
 		let handle = PeersetHandle {
 			tx,
 		};
 
 		let mut peerset = Peerset {
-			data,
+			data: peersstate::PeersState::new(config.in_peers, config.out_peers),
 			rx,
+			reserved_only: config.reserved_only,
 			message_queue: VecDeque::new(),
+			created: Instant::now(),
+			latest_time_update: Instant::now(),
 		};
 
 		for peer_id in config.reserved_nodes {
-			peerset.data.discovered.add_peer(peer_id, SlotType::Reserved);
+			if let peersstate::Peer::Unknown(entry) = peerset.data.peer(&peer_id) {
+				entry.discover().set_reserved(true);
+			} else {
+				debug!(target: "peerset", "Duplicate reserved node in config: {:?}", peer_id);
+			}
 		}
 
 		for peer_id in config.bootnodes {
-			peerset.data.discovered.add_peer(peer_id, SlotType::Common);
+			if let peersstate::Peer::Unknown(entry) = peerset.data.peer(&peer_id) {
+				entry.discover();
+			} else {
+				debug!(target: "peerset", "Duplicate bootnode in config: {:?}", peer_id);
+			}
 		}
 
 		peerset.alloc_slots();
 		(peerset, handle)
 	}
 
 	fn on_add_reserved_peer(&mut self, peer_id: PeerId) {
-		// Nothing more to do if we're already connected.
-		if self.data.in_slots.contains(&peer_id) {
-			self.data.in_slots.mark_reserved(&peer_id);
-			return;
-		}
-
-		match self.data.out_slots.add_peer(peer_id, SlotType::Reserved) {
-			SlotState::Added(peer_id) => {
-				// reserved node may have been previously stored as normal node in discovered list
-				self.data.discovered.remove_peer(&peer_id);
-
-				// notify that connection has been made
-				trace!(target: "peerset", "Connecting to new reserved peer {}", peer_id);
-				self.message_queue.push_back(Message::Connect(peer_id));
-				return;
-			},
-			SlotState::Swaped { removed, added } => {
-				// reserved node may have been previously stored as normal node in discovered list
-				self.data.discovered.remove_peer(&added);
-				// let's add the peer we disconnected from to the discovered list again
-				self.data.discovered.add_peer(removed.clone(), SlotType::Common);
-				// swap connections
-				trace!(target: "peerset", "Connecting to new reserved peer {}, dropping {}", added, removed);
-				self.message_queue.push_back(Message::Drop(removed));
-				self.message_queue.push_back(Message::Connect(added));
-			}
-			SlotState::AlreadyExists(_) | SlotState::Upgraded(_) => {
-				return;
-			}
-			SlotState::MaxCapacity(peer_id) => {
-				self.data.discovered.add_peer(peer_id, SlotType::Reserved);
-				return;
+		let mut entry = match self.data.peer(&peer_id) {
```

Comment:
A general suggestion on the `PeerState` design:

Have you considered just keeping the entire data hidden inside of it? For example here, why not just do a `let connected = self.data.add_reserved_peer(peer_id);`, followed by `if connected {// queue the message}`. 

It appears to me that almost every operation made on `self.data` in this file, which currently involves making the internal data of the `PeerState` available to the client for manipulation, could be replaced by a method call on the `PeerState`, with a return value if necessary. 

The current design seems to introduce a new data structure, the `PeerState`, and then make it's internal available to the client. My suggestion would be to either not introduce a new data structure, and just put everything right on the `Peerset`, or hide the internal state of `PeerState`. Either way you'd be able to simplify things, and remove what is currently done with `Peer`. 

Just to drive the point home, currently inside the loop at `alloc_slots`, we see the following:

1. `self.data.reserved_not_connected_peer()`
2. `self.data.highest_not_connected_peer()`
3. `next.reputation()` (with `next` being a variant of `Peer`)
4. `match next.try_outgoing()`, 
5. `Ok(conn) => self.message_queue.push_back(Message::Connect(conn.into_peer_id())),`

What's the usefulness of `PeerState` as an abstraction if the client needs to do all this stuff?

I would suggest something more like: 

```rust
match self.data.next_peer_to_connect_to() {
    Some(peer_id) => // enqueue the Connect message
    None => // No slot available
}
```
or, just put everything directly on the peerset, with no need for `PeerState`.

Reply:
The point is to separate the data structure from the logic. The logic right now is extremely simple, but once it becomes more complex it's IMO better to not have to deal with the increased complexity of both deciding who to connect to and how to store data in memory.

Reply:
Ok I see, in that case it does make sense to share `Peer` as a handle to keep the business logic in the peerset.